src/number_theory/sum_four_squares.lean

```diff
@@ -77,16 +80,62 @@ open int
 
 open_locale classical
 
+@[simp] lemma zmod_two_sq_eq_self (a : zmod 2) : a ^ 2 = a :=
+by change fin 2 at a; fin_cases a; simp
+
+@[simp] lemma zmod_two_two_eq_zero : (2 : zmod 2) = 0 :=
+suffices ((2 : ℕ) : zmod 2) = (0 : ℕ), by norm_num at this; assumption,
+by rw [zmod.nat_coe_eq_nat_coe_iff, nat.modeq]; norm_num
+
+private lemma forall_fin_succ_iff {α} {n} (p : (fin (n+1) → α) → Prop) :
+  (∀ x, p x) ↔ ∀ x a, p (matrix.vec_cons a x) :=
+⟨λ h, by simp *, λ h x, by { rw [← matrix.cons_head_tail x], apply h }⟩
+
+private lemma forall_fin_zero_iff {α} (p : (fin 0 → α) → Prop) :
+  (∀ x, p x) ↔ p matrix.vec_empty :=
+⟨λ h, by simp *, λ h x, by cc⟩
+
+private lemma forall_zmod_two_iff (p : zmod 2 → Prop) : (∀ a, p a) ↔ p 0 ∧ p 1 :=
+⟨λ h, by simp *, λ ⟨_,_⟩ (a : fin 2), by fin_cases a; assumption⟩
+
+private lemma exists_fin_succ_iff {n} (p : fin (n+1) → Prop) :
+  (∃ i, p i) ↔ p 0 ∨ ∃ i : fin n, p i.succ :=
+⟨begin
+  rintros ⟨⟨_|i, hi⟩, h⟩,
+  { left, assumption },
+  { right, refine ⟨⟨i, succ_lt_succ_iff.1 hi⟩, _⟩, assumption },
+ end,
+ by rintros (h|⟨_,h⟩); exact ⟨_, h⟩⟩
+
+private lemma exists_fin_zero_iff (p : fin 0 → Prop) : (∃ i, p i) ↔ false :=
+⟨λ ⟨⟨i, hi⟩, h⟩, by cases hi, by simp⟩
+
+private lemma two_eq_parity_of_sum_four_squares :
+  ∀ f : fin 4 → zmod 2, (f 0)^2 + (f 1)^2 + (f 2)^2 + (f 3)^2 = 0 →
+    ∃ i : (fin 4), (f i)^2 + f (swap i 0 1)^2 = 0 ∧ f (swap i 0 2)^2 + f (swap i 0 3)^2 = 0 :=
+begin
```

Comment:
This used to be `dec_trivial`, but now statements involving `nat.mod` no longer reduce in the kernel.

I can see two ways out here:
 - Write automation to prove this kind of statements, i.e. `norm_num++^🍀`, if you will.  This has the advantage of using the asymptotically faster algorithms in `norm_num`.
 - Define `nat.mod` without well-founded recursion.